dsl/flatten.cc

```diff
@@ -0,0 +1,304 @@
+// correct. Finally, because methods evaluate to their corresponding symbols, the former location of `bar`'s definition
+// has been replaced with the verbatim symbol `:bar`.
+//
+// It does this by maintaining both a stack and a queue: elements on the stack include indices into the queue as well as
+// scoping information, while the queue consists of pointers to expressions; additionally, some stack elements do not
+// correspond to expressions that need to be moved, in which case their indices will be -1: for example, all methods at
+// the root of a class definition will have stack entries (so we can properly track nested methods) but none of them
+// will need to be moved, so they will have -1 indices.
+namespace sorbet::dsl {
+
+class FlattenWalk {
+    struct MethodData {
+        // This is an index into the methods stack
+        int idx;
+        // we need to keep information around about whether we're in a static outer context: for example, if we have
+        //
+        //   def self.foo; def bar; end; end
+        //
+        // then we should flatten it to
+        //
+        //  def self.foo; end
+        //  def self.bar; end
+        //
+        // which means when we get to `bar` we need to know that the outer context `foo` is static. We pass that down
+        // the current stack by means of this `isStatic` variable.
+        int staticLevel;
+        ast::Expression *target;
+        MethodData(int idx, int staticLevel, ast::Expression *target)
+            : idx(idx), staticLevel(staticLevel), target(target){};
+    };
+    struct MovedItem {
+        unique_ptr<ast::Expression> expr;
+        int staticLevel;
+        MovedItem(unique_ptr<ast::Expression> expr, int staticLevel) : expr(move(expr)), staticLevel(staticLevel){};
+        MovedItem() = default;
+    };
+    struct Methods {
+        vector<MovedItem> methods;
+        vector<MethodData> stack;
+        Methods() = default;
+    };
+
+    vector<Methods> methodScopes;
+
+    int computeStaticLevel(const ast::MethodDef &methodDef) {
+        auto &methods = curMethodSet();
+        int prevLevel = methods.stack.empty() ? 0 : methods.stack.back().staticLevel;
+        return prevLevel + (methodDef.isSelf() ? 1 : 0);
+    }
+
+    void newMethodSet() {
+        methodScopes.emplace_back();
+    }
+    Methods &curMethodSet() {
+        ENFORCE(!methodScopes.empty());
+        return methodScopes.back();
+    }
+    void popCurMethodSet() {
+        ENFORCE(!methodScopes.empty());
+        methodScopes.pop_back();
+    }
+
+    vector<MovedItem> popCurMethodDefs() {
+        auto ret = std::move(curMethodSet().methods);
+        ENFORCE(curMethodSet().stack.empty());
+        popCurMethodSet();
+        return ret;
+    };
+
+    // extract all the methods from the current queue and put them at the end of the class's current body
+    ast::ClassDef::RHS_store addMethods(core::Context ctx, ast::ClassDef::RHS_store rhs, core::Loc loc) {
```

Comment:
Do you have two different methods called `addMethods` in this class?

Reply:
I think the only reason why the code currently works is because those methods have different signatures, so C++ overload resolution kicks in... Can we rename one or both of them?